File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/mailbox/v2/MailboxV2Dtos.kt
package app.pantopus.android.data.api.models.mailbox.v2

import app.pantopus.android.data.api.models.common.JsonArrayValue
import app.pantopus.android.data.api.models.common.JsonValue
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/** `GET /api/mailbox/v2/drawers` — route `backend/routes/mailboxV2.js:214`. */
@JsonClass(generateAdapter = true)
data class DrawerListResponse(
    val drawers: List<Drawer>,
)

@JsonClass(generateAdapter = true)
data class Drawer(
    val drawer: String,
    @Json(name = "display_name") val displayName: String,
    val icon: String,
    @Json(name = "unread_count") val unreadCount: Int,
    @Json(name = "urgent_count") val urgentCount: Int,
    @Json(name = "last_item_at") val lastItemAt: String?,
)

/** `GET /api/mailbox/v2/drawer/:drawer` — route `backend/routes/mailboxV2.js:280`. */
@JsonClass(generateAdapter = true)
data class DrawerItemsResponse(
    val mail: List<DrawerMail>,
    val total: Int,
    val drawer: String,
)

@JsonClass(generateAdapter = true)
data class DrawerMail(
    val id: String,
    val type: String,
    @Json(name = "mail_type") val mailType: String?,
    val subject: String?,
    @Json(name = "created_at") val createdAt: String,
    @Json(name = "display_title") val displayTitle: String?,
    @Json(name = "preview_text") val previewText: String?,
    @Json(name = "ack_required") val ackRequired: Boolean?,
    @Json(name = "ack_status") val ackStatus: String?,
    val viewed: Boolean = false,
    val starred: Boolean = false,
    val archived: Boolean = false,
    val sender: SenderRef?,
    @Json(name = "sender_business_name") val senderBusinessName: String?,
    @Json(name = "sender_address") val senderAddress: String?,
    @Json(name = "sender_display") val senderDisplay: String,
    @Json(name = "sender_trust") val senderTrust: String,
    val `package`: JsonValue?,
) {
    @JsonClass(generateAdapter = true)
    data class SenderRef(
        val name: String?,
        val username: String?,
    )
}

/** `GET /api/mailbox/v2/item/:id` — route `backend/routes/mailboxV2.js:366`. */
@JsonClass(generateAdapter = true)
data class MailboxV2ItemResponse(
    val mail: MailboxV2Item,
)

@JsonClass(generateAdapter = true)
data class MailboxV2Item(
    val id: String,
    val type: String,
    @Json(name = "created_at") val createdAt: String,
    @Json(name = "display_title") val displayTitle: String?,
    @Json(name = "preview_text") val previewText: String?,
    val viewed: Boolean = false,
    val archived: Boolean = false,
    val sender: DrawerMail.SenderRef?,
    @Json(name = "sender_display") val senderDisplay: String,
    @Json(name = "sender_trust") val senderTrust: String,
    /**
     * Pantopus user id behind the sender. Sourced from the `*`
     * projection on Mail in `backend/routes/mailboxV2.js`. Drives the
     * avatar-tap → public profile navigation in P17.
     */
    @Json(name = "sender_user_id") val senderUserId: String? = null,
    val `package`: JsonValue?,
    val packageInfo: JsonValue? = null,
    @Json(name = "timeline") val packageTimeline: JsonArrayValue = emptyList(),
    @Json(name = "object_payload") val objectPayload: JsonValue? = null,
    /** Raw `Mail.subject` column. Surfaced so T3.8 Ceremonial Mail
     *  Open can render the letter heading; v1 mail-detail clients
     *  already had this. */
    val subject: String? = null,
    /** Raw `Mail.content` column for the letter body. */
    val content: String? = null,
    @Json(name = "action_required") val actionRequired: Boolean? = null,
    @Json(name = "ack_required") val ackRequired: Boolean? = null,
    val tags: List<String> = emptyList(),
    val attachments: List<String>? = null,
)

/** `POST /api/mailbox/v2/item/:id/action` — route `backend/routes/mailboxV2.js:459`. */
@JsonClass(generateAdapter = true)
data class MailboxItemActionRequest(
    /** One of: pay, sign, forward, file, shred, remind, split, acknowledge, share_household, create_task, dispute. */
    val action: String,
)

@JsonClass(generateAdapter = true)
data class MailboxItemActionResponse(
    val message: String,
    val action: String,
)

/**
 * `GET /api/mailbox/v2/package/:mailId` — route `backend/routes/mailboxV2.js:634`.
 * Package + timeline columns evolve server-side; they're exposed as JSON
 * maps rather than invented types.
 */
@JsonClass(generateAdapter = true)
data class PackageDetailResponse(
    val `package`: JsonValue,
    val timeline: JsonArrayValue,
    val sender: Sender?,
) {
    @JsonClass(generateAdapter = true)
    data class Sender(
        val display: String,
        val trust: String,
    )
}

/** `PATCH /api/mailbox/v2/package/:mailId/status` — route `backend/routes/mailboxV2.js:670`. */
@JsonClass(generateAdapter = true)
data class PackageStatusUpdateRequest(
    /** One of: pre_receipt, in_transit, out_for_delivery, delivered, exception. */
    val status: String,
    val location: String? = null,
    val photoUrl: String? = null,
    val deliveryLocationNote: String? = null,
)

@JsonClass(generateAdapter = true)
data class PackageStatusUpdateResponse(
    val message: String,
    val status: String,
    val previousStatus: String,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Mailbox/V2/MailboxV2DTOs.swift
//
//  MailboxV2DTOs.swift
//  Pantopus
//
//  DTOs for the V2 mailbox endpoints in `backend/routes/mailboxV2.js`.
//

import Foundation

/// `GET /api/mailbox/v2/drawers` envelope — route `backend/routes/mailboxV2.js:214`.
public struct DrawerListResponse: Decodable, Sendable, Hashable {
    public let drawers: [Drawer]

    public struct Drawer: Decodable, Sendable, Hashable, Identifiable {
        public let drawer: String
        public let displayName: String
        public let icon: String
        public let unreadCount: Int
        public let urgentCount: Int
        public let lastItemAt: String?

        public var id: String {
            drawer
        }

        private enum CodingKeys: String, CodingKey {
            case drawer
            case displayName = "display_name"
            case icon
            case unreadCount = "unread_count"
            case urgentCount = "urgent_count"
            case lastItemAt = "last_item_at"
        }
    }
}

/// `GET /api/mailbox/v2/drawer/:drawer` envelope — route `backend/routes/mailboxV2.js:280`.
public struct DrawerItemsResponse: Decodable, Sendable, Hashable {
    public let mail: [DrawerMail]
    public let total: Int
    public let drawer: String

    public struct DrawerMail: Decodable, Sendable, Hashable, Identifiable {
        public let item: MailItem
        public let sender: SenderRef?
        public let senderDisplay: String
        public let senderTrust: String
        public let package: JSONValue?

        public var id: String {
            item.id
        }

        public init(from decoder: any Decoder) throws {
            item = try MailItem(from: decoder)
            let c = try decoder.container(keyedBy: Keys.self)
            sender = try c.decodeIfPresent(SenderRef.self, forKey: .sender)
            senderDisplay = try c.decode(String.self, forKey: .senderDisplay)
            senderTrust = try c.decode(String.self, forKey: .senderTrust)
            package = try c.decodeIfPresent(JSONValue.self, forKey: .package)
        }

        private enum Keys: String, CodingKey {
            case sender
            case senderDisplay = "sender_display"
            case senderTrust = "sender_trust"
            case package
        }

        public struct SenderRef: Decodable, Sendable, Hashable {
            public let name: String?
            public let username: String?
        }
    }
}

/// `GET /api/mailbox/v2/item/:id` envelope — route `backend/routes/mailboxV2.js:366`.
public struct MailboxV2ItemResponse: Decodable, Sendable, Hashable {
    public let mail: Item

    public struct Item: Decodable, Sendable, Hashable, Identifiable {
        public let base: MailItem
        public let sender: DrawerItemsResponse.DrawerMail.SenderRef?
        public let senderDisplay: String
        public let senderTrust: String
        public let package: JSONValue?
        public let packageInfo: JSONValue?
        public let packageTimeline: [JSONValue]
        public let objectPayload: JSONValue?

        public var id: String {
            base.id
        }

        public init(from decoder: any Decoder) throws {
            base = try MailItem(from: decoder)
            let c = try decoder.container(keyedBy: Keys.self)
            sender = try c.decodeIfPresent(DrawerItemsResponse.DrawerMail.SenderRef.self, forKey: .sender)
            senderDisplay = try c.decode(String.self, forKey: .senderDisplay)
            senderTrust = try c.decode(String.self, forKey: .senderTrust)
            package = try c.decodeIfPresent(JSONValue.self, forKey: .package)
            packageInfo = try c.decodeIfPresent(JSONValue.self, forKey: .packageInfo)
            packageTimeline = try c.decodeIfPresent([JSONValue].self, forKey: .packageTimeline) ?? []
            objectPayload = try c.decodeIfPresent(JSONValue.self, forKey: .objectPayload)
        }

        private enum Keys: String, CodingKey {
            case sender
            case senderDisplay = "sender_display"
            case senderTrust = "sender_trust"
            case package
            case packageInfo // backend: column on Mail (rare); currently no override
            case packageTimeline = "timeline" // backend response key
            case objectPayload = "object_payload" // backend response key (P18 fix)
        }
    }
}

/// `POST /api/mailbox/v2/item/:id/action` request body. Route:
/// `backend/routes/mailboxV2.js:459`.
public struct MailboxItemActionRequest: Encodable, Sendable {
    /// One of: `pay`, `sign`, `forward`, `file`, `shred`, `remind`, `split`,
    /// `acknowledge`, `share_household`, `create_task`, `dispute`.
    public let action: String

    public init(action: String) {
        self.action = action
    }
}

/// `POST /api/mailbox/v2/item/:id/action` response.
public struct MailboxItemActionResponse: Decodable, Sendable, Hashable {
    public let message: String
    public let action: String
}

/// `GET /api/mailbox/v2/package/:mailId` envelope — route
/// `backend/routes/mailboxV2.js:634`. The `package` and `timeline` elements
/// are rich rows whose columns evolve; they're exposed as `JSONValue` to
/// avoid drift between the server schema and the mobile clients.
public struct PackageDetailResponse: Decodable, Sendable, Hashable {
    public let package: JSONValue
    public let timeline: [JSONValue]
    public let sender: Sender?

    public struct Sender: Decodable, Sendable, Hashable {
        public let display: String
        public let trust: String
    }
}

/// `PATCH /api/mailbox/v2/package/:mailId/status` request.
public struct PackageStatusUpdateRequest: Encodable, Sendable {
    /// One of: `pre_receipt`, `in_transit`, `out_for_delivery`,
    /// `delivered`, `exception`.
    public let status: String
    public let location: String?
    public let photoUrl: String?
    public let deliveryLocationNote: String?

    public init(
        status: String,
        location: String? = nil,
        photoUrl: String? = nil,
        deliveryLocationNote: String? = nil
    ) {
        self.status = status
        self.location = location
        self.photoUrl = photoUrl
        self.deliveryLocationNote = deliveryLocationNote
    }
}

/// `PATCH /api/mailbox/v2/package/:mailId/status` response.
public struct PackageStatusUpdateResponse: Decodable, Sendable, Hashable {
    public let message: String
    public let status: String
    public let previousStatus: String
}
